From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: EJ Hsu <ejh@nvidia.com>
Date: Mon, 12 Aug 2024 09:02:44 +0000
Subject: [PATCH 058/254] NVIDIA: SAUCE: phy: tegra: fix memory corruption

In XUSB virtualization, the padctl interrupt will be directed to
xhci server. In order for VF to get notified when padctl interrupt
is triggered, VF needs to register a notifier to padctl driver during
its probe().

Add tegra_xusb_padctl_event_unregister() so that VF can remove its
notifier when unbind operation is called.

http://nvbugs/4762786

Change-Id: I2878cf1ae062d602314123baae07d4ddbb8fbb3f
Signed-off-by: EJ Hsu <ejh@nvidia.com>
(cherry picked from commit de84840f20006929879297fcd467f3c66442649a)
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3201123
GVS: buildbot_gerritrpt <buildbot_gerritrpt@nvidia.com>
Tested-by: Petlozu Pravareshwar <petlozup@nvidia.com>
Reviewed-by: Laxman Dewangan <ldewangan@nvidia.com>
Reviewed-by: Petlozu Pravareshwar <petlozup@nvidia.com>

diff --git a/drivers/phy/tegra/xusb.c b/drivers/phy/tegra/xusb.c
index f7bd83e5f3504..e85341bdd5492 100644
--- a/drivers/phy/tegra/xusb.c
+++ b/drivers/phy/tegra/xusb.c
@@ -1388,6 +1388,13 @@ int tegra_xusb_padctl_event_register(struct tegra_xusb_padctl *padctl,
 }
 EXPORT_SYMBOL_GPL(tegra_xusb_padctl_event_register);
 
+int tegra_xusb_padctl_event_unregister(struct tegra_xusb_padctl *padctl,
+				     struct notifier_block *nb)
+{
+	return blocking_notifier_chain_unregister(&padctl->notifier, nb);
+}
+EXPORT_SYMBOL_GPL(tegra_xusb_padctl_event_unregister);
+
 int tegra_xusb_padctl_event_notify(struct tegra_xusb_padctl *padctl,
 				     unsigned long val)
 {
diff --git a/include/linux/phy/tegra/xusb.h b/include/linux/phy/tegra/xusb.h
index bcb0e9ee3d468..e547490b33975 100644
--- a/include/linux/phy/tegra/xusb.h
+++ b/include/linux/phy/tegra/xusb.h
@@ -1,6 +1,6 @@
 /* SPDX-License-Identifier: GPL-2.0-only */
 /*
- * Copyright (c) 2016-2022, NVIDIA CORPORATION.  All rights reserved.
+ * SPDX-FileCopyrightText: Copyright (c) 2016-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
  */
 
 #ifndef PHY_TEGRA_XUSB_H
@@ -20,6 +20,8 @@ struct tegra_xusb_padctl *tegra_xusb_padctl_get(struct device *dev);
 void tegra_xusb_padctl_put(struct tegra_xusb_padctl *padctl);
 int tegra_xusb_padctl_event_register(struct tegra_xusb_padctl *padctl,
 				     struct notifier_block *nb);
+int tegra_xusb_padctl_event_unregister(struct tegra_xusb_padctl *padctl,
+				     struct notifier_block *nb);
 int tegra_xusb_padctl_event_notify(struct tegra_xusb_padctl *padctl,
 				     unsigned long val);
 int tegra_xusb_padctl_usb3_save_context(struct tegra_xusb_padctl *padctl,
-- 
2.43.0

